Added Client.customerSubscriptions

stevensnoeijen opened the pull request:
Added concept for Client.customerSubscriptions to create, getById and delete (cancel) subscriptions with tests.

For this I also added:

- Selenium to fully automatic test a payment. I needed this to able to test the creation of a subscription (which requires a first payment).

- DeleteWithResponse, due to the Delete-concept would not resolve an response object. I needed this because the cancelation of a customerSubscription is done by a DELETE method that have a response when the enddate is.

Commit:
added: CustomerPayments added GetById

Commit:
added Client.customerSubscriptions

added: DeleteWithResponse concept
added: Client.customerSubscriptions with create, getById and delete with tests
added: selenium for test in CustomerSubscriptionsIntegrationTest

Commit:
changed selenium to be installed by webdrivermanager

Commit:
changed: CreateSubscription.startDate changed to String, due this gave problems with jsonformat

Commit:
added: test for customersubscription cancel

stevensnoeijen commented:
looks like Travis dont like that I'm using selenium, ill check this soon

stil4m commented on src/main/java/nl/stil4m/mollie/Client.java:
```diff
@@ -1,21 +1,12 @@
-import nl.stil4m.mollie.concepts.Refunds;
-import nl.stil4m.mollie.concepts.PaymentRefunds;
-import nl.stil4m.mollie.concepts.Status;
+import nl.stil4m.mollie.concepts.*;
```
Please do not use `*` imports

stil4m commented on src/main/java/nl/stil4m/mollie/domain/CreateSubscription.java:
```diff
@@ -0,0 +1,94 @@
+    private Optional<String> webhookUrl;
+
+    public CreateSubscription(@Nonnull Double amount, @Nullable Integer times, @Nonnull String interval, @Nullable String startDate, @Nonnull String description, @Nullable String method, @Nullable String webhookUrl) {
+        super();
```
Is this `super()` call required?

stil4m commented on src/test/java/nl/stil4m/mollie/concepts/CustomerSubscriptionsIntegrationTest.java:
```diff
@@ -0,0 +1,176 @@
+        //open paymentUrl and set it to paid
+
+        //setup selenium
+        WebDriver driver = new ChromeDriver();
```
Shouldn't this belong to the before?

stil4m commented on src/test/java/nl/stil4m/mollie/concepts/CustomerSubscriptionsIntegrationTest.java:
```diff
@@ -0,0 +1,176 @@
+
+        //check if payment is complete
+        payment = customerPayments.get(payment.getId()).getData();//get new version of payment
+        if (!payment.getStatus().equals("paid")) {
```
Would it be a good interface to add a `isPaid()` method to payment? (which will be ignored in the JSON serialization)

stevensnoeijen replied:
I don't agree to just add `isPaid()`, because a payment can have 9 types of statuses (source:  https://www.mollie.com/en/docs/status)
I would suggest making it an enum, what do you think about that?

stevensnoeijen commented:
I'm quite bussy atm, I will react on this in a few weeks...

Commit:
processed comments on pull request https://github.com/stil4m/mollie-api/pull/51

added project's nb-configuration for netbeans so that imports wont *
fixed imports *
removed unneeded super()
moved Selenium Webdriver setup to @BeforeClass and @Before

stevensnoeijen commented:
Ok, i've processed your requests expect one, please see my answer above. 
Travis still dont like the selenium driver, can you look at this https://docs.travis-ci.com/user/gui-and-headless-browsers/#Using-the-Chrome-addon-in-the-headless-modemaybe this can fix that?

Commit:
updated selenium webdrivermanager lib
categorised all libs by test

stevensnoeijen commented:
bump

Commit:
changed: Payment and Refund datetime's to Date objects instead of strings
changed: added Payment.failedDatetime
changed: JsonIgnoreProperies(ignoreUnknown = true) for payment and Refund
fixed: ignore IssuersInegrationTest by ignoring the current failing test, due that issuers where changed?
added: test for PaymentIntegrationTest by payment.status paid, cancelled, refunded, expired
fixed: RefundIntegerationTest due that there is a refund payment it wont return 0 results anymore

Commit:
added: Payment.getFailedDatetime

stil4m commented on nb-configuration.xml:
```diff
@@ -0,0 +1,26 @@
+        <org-netbeans-modules-editor-indent.CodeStyle.usedProfile>project</org-netbeans-modules-editor-indent.CodeStyle.usedProfile>
+        <org-netbeans-modules-javascript2-requirejs.enabled>true</org-netbeans-modules-javascript2-requirejs.enabled>
+    </properties>
+</project-shared-configuration>
```
Please do not checkin this file

stevensnoeijen replied:
gitignore-ed this file

stil4m commented on src/main/java/nl/stil4m/mollie/DynamicClient.java:
```diff
@@ -1,14 +1,6 @@
-import nl.stil4m.mollie.concepts.Refunds;
-import nl.stil4m.mollie.concepts.PaymentRefunds;
-import nl.stil4m.mollie.concepts.Status;
+import nl.stil4m.mollie.concepts.*;
```
Please revert this

stil4m commented on src/main/java/nl/stil4m/mollie/domain/Payment.java:
```diff
@@ -36,29 +36,30 @@
+            @JsonProperty("locale") String locale,
+            @JsonProperty("countryCode") String countryCode,
+            @JsonProperty("expiryPeriod") String expiryPeriod,
+            @JsonProperty("mandateId") Optional<String> mandateId,
```
Please indent vertically with first parameter

stevensnoeijen replied:
I dont understand, can you make an example?

stil4m commented on src/main/java/nl/stil4m/mollie/domain/Payment.java:
```diff
@@ -18,9 +17,10 @@
-    private final String expiredDatetime;
+    private final Date paidDatetime;
+    private final Date cancelledDatetime;
+    private final Date expiredDatetime;
```
Like :)

stevensnoeijen replied:
yay :)

stil4m commented on src/main/java/nl/stil4m/mollie/domain/Payment.java:
```diff
@@ -36,29 +36,30 @@
+            @JsonProperty("expiryPeriod") String expiryPeriod,
+            @JsonProperty("mandateId") Optional<String> mandateId,
+            @JsonProperty("canBeCancelled") Optional<Boolean> canBeCancelled,
+            @JsonProperty("failedDatetime") Date failedDatetime) {
```
I only sea `failedAt` in the API (https://docs.mollie.com/reference/v2/payments-api/get-payment). Where does this come from?

stevensnoeijen replied:
you're looking at v2, this version recently released
check https://docs.mollie.com/reference/v1/payments-api/get-payment

stil4m commented on src/main/java/nl/stil4m/mollie/domain/Refund.java:
```diff
@@ -21,11 +22,11 @@
+            @JsonProperty("amount") Double amount,
+            @JsonProperty("status") String status,
+            @JsonProperty("refundedDatetime") Date refundedDatetime,
+            @JsonProperty("links") RefundLinks links) {
```
idem

stil4m commented on src/main/java/nl/stil4m/mollie/domain/Subscription.java:
```diff
@@ -0,0 +1,179 @@
+/**
+ * @see https://www.mollie.com/en/docs/reference/subscriptions/get
+ * @author stevensnoeijen
+ */
```
Please remove the author. We have Git for this.

stevensnoeijen replied:
removed @author from code

stil4m commented on src/test/java/nl/stil4m/mollie/concepts/PaymentsIntegrationTest.java:
```diff
@@ -1,35 +1,33 @@
+import nl.stil4m.mollie.domain.Payment;
+import nl.stil4m.mollie.domain.subpayments.ideal.CreateIdealPayment;
+import nl.stil4m.mollie.domain.subpayments.ideal.IdealPaymentOptions;
+import org.apache.http.HttpStatus;
```
It seems that your import order settings are not set up the same. Can you layout static imports separately?

My setup is:
```
com.*

nl.*

org.*

javax.*

java.*

other imports

other static imports

```

stevensnoeijen replied:
i've changed to intellij now and set the project import settings to match this, it should match it now :)

Commit:
added: changeback in client by paymentId and without paymentId

Commit:
added pull change requests https://github.com/stil4m/mollie-api/pull/51

added: gitignore nb-configuration.xml (netbeans)
changed: import in DynamicClient
changed: removed @author from javadoc

Commit:
changed imports order for https://github.com/stil4m/mollie-api/pull/51

changed: import order to match:
com.*

nl.*

org.*

javax.*

java.*

other imports

other static imports

stevensnoeijen commented:
In the last week I also added the changeback api in client.paymentChargebacks and client.chargebacks, but without tests, because this isnt testable with the current mollie test api.

Commit:
Possible fix for updating customer

Commit:
Fixed issue with updating a customer

Commit:
added subscriptionId to Payment

stevensnoeijen commented:
bump.
But test (IT) fail because mollie changed their response somewhat and some fail because some id's (like payment) are probably cleaned up. And Travis dont like selenium tests.
What would you suggest?
